---
title: "Scseq_preprocessing_d3"
author: "Khang"
date: "r Sys.Date()"
output:
  html_document:
    df_print: paged
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE,message = FALSE,warning = FALSE)
knitr::opts_chunk$set(dev = "pdf",
                      dpi = 600,
                      cache = TRUE)

library(immunarch)
library(tidyverse)
library(RColorBrewer)
library(cowplot)
library(seqinr)
```

```{r load_data}
files_mixcr <- list.files("./1_Ig_all_unfiltered_contigs/", full.names = T)
sample_all <- map(.x = files_mixcr, function (x) read_tsv(x))
names(sample_all) <- str_split(list.files("./1_Ig_all_unfiltered_contigs/"),"\\.", simplify = T)[,1]
count_mixcr_out <- tibble(Sample = names(sample_all), 
                          Count = map(sample_all, function(x) nrow(x)) %>% unlist,
                          Type = "MiXCR_out")
```

# Check if all contigs have CDR3 and defined isotype (no missing aaseqCDR3, bestVGene, bestJGene, or bestCGene)

```{r Contigs_with_CDR3_isotype, fig.height=5, fig.width=8}
sample_all <- map(sample_all, function (x)
  filter(.data = x, !is.na(aaSeqCDR3) & !is.na(bestVGene) & !is.na(bestJGene) & !is.na(bestCGene)))

count_sample_all <- tibble(Sample = names(sample_all), 
                           Count = map(sample_all, function(x) nrow(x)) %>% unlist,
                           Type = "Defined_isotype")
ggplot(count_sample_all) +
  geom_col(aes(x = Sample, y = Count, fill = Sample), color = "black") +
  scale_fill_brewer(type = "qual", palette = "Set3") +
  theme_half_open() +
  theme(legend.position = "right", axis.text.x = element_text(angle = 90)) +
  geom_text(aes(x = Sample, y = Count, label = Count), vjust = -0.2) +
  labs(title = "Contigs with CDR3 and defined isotype")
```

# Filter for singlet droplets (1 cell barcode appear only twice)

```{r contigs_singlets, fig.height=5, fig.width=8}
# List out barcode_two in all samples
barcode_two <- sapply(sample_all, function(x) names(which((table(x$tagValueCELL)) == "2")))

# Select for rows that have barcode_two
#sample_1[which(sample_1$barcode %in% barcode_two_all$sample_1),]
sample_all_paired <- map2(.x = sample_all, .y = barcode_two, function (x,y) x[which(x$tagValueCELL %in% y),])

count_sample_all_paired <- tibble(Sample = names(sample_all_paired), 
                           Count = map(sample_all_paired, function(x) nrow(x)) %>% unlist,
                           Type = "Singlets")

ggplot(count_sample_all_paired) +
  geom_col(aes(x = Sample, y = Count, fill = Sample), color = "black") +
  scale_fill_brewer(type = "qual", palette = "Set3") +
  theme_half_open() +
  theme(legend.position = "right", axis.text.x = element_text(angle = 90)) +
  geom_text(aes(x = Sample, y = Count, label = Count), vjust = -0.2) +
  labs(title = "Contigs with 2 barcodes (single cell droplets)")
```

# Check if the barcode pairing is between one heavy and one light chain and filter

```{r contigs_paired_correct, fig.height=5, fig.width=8}
# Group the df by cell barcode and paste the chain name together
sample_all_paired_2 <- map(sample_all_paired, function(x) 
                           x %>% group_by(tagValueCELL) %>% 
                           mutate(chainsPair = paste(chains, collapse = ";")))

# Filter the df by allowing only heavy and light chain pairing
sample_all_paired_2 <- map(sample_all_paired_2, function(x)
                           filter(x, chainsPair == "IGH;IGK"| chainsPair == "IGH;IGL"|
                                  chainsPair == "IGK;IGH"| chainsPair == "IGL;IGH"))
# Double check the pairing
lapply(sample_all_paired_2, function(x) table(x$chainsPair))

count_sample_all_paired_2 <- tibble(Sample = names(sample_all_paired_2), 
                           Count = map(sample_all_paired_2, function(x) nrow(x)) %>% unlist,
                           Type = "Correct_pairing")
ggplot(count_sample_all_paired_2) +
  geom_col(aes(x = Sample, y = Count, fill = Sample), color = "black") +
  scale_fill_brewer(type = "qual", palette = "Set3") +
  theme_half_open() +
  theme(legend.position = "right", axis.text.x = element_text(angle = 90)) +
  geom_text(aes(x = Sample, y = Count, label = Count), vjust = -0.2) +
  labs(title = "Contigs with correct chain pairing (HC/LC)")
```

# Save the contig outputs
```{r}
# Save the filtered contigs
#map2(.x = sample_all_paired_2, .y = names(sample_all_paired_2), function (x,y) write_tsv(x, file = paste0("./2_Ig_all_filtered_contigs/",y,".txt")))
```

# Track how much contigs are lost after all pre-processing steps
```{r contigs_track, fig.height=8, fig.width=8}
count_processing <- rbind(count_mixcr_out, count_sample_all, count_sample_all_paired,
                          count_sample_all_paired_2)
count_processing <- count_processing %>% group_by(Sample) %>% mutate(Proportion = Count/max(Count))

ggplot(count_processing) +
  geom_step(aes(x = factor(Type, levels = c("MiXCR_out", "Defined_isotype", "Singlets", "Correct_pairing")), y = Count, group = Sample, color = Sample)) +
  geom_point(aes(x = factor(Type, levels = c("MiXCR_out", "Defined_isotype", "Singlets", "Correct_pairing")), y = Count, group = Sample, color = Sample)) +
  geom_text(aes(x = factor(Type, levels = c("MiXCR_out", "Defined_isotype", "Singlets", "Correct_pairing")), y = Count, group = Sample, color = Sample, label = Count), hjust = 1.2, check_overlap = TRUE) +
  scale_color_brewer(type = "qual", palette = "Paired") +
  theme_half_open() +
  theme(legend.position = "right", axis.text.x = element_text(angle = 0)) +
  labs(title = "Contigs remaining after processing", x = "Processing step")
```
\newpage
```{r contigs_track_prop, fig.height=8, fig.width=8}
ggplot(count_processing) +
  geom_step(aes(x = factor(Type, levels = c("MiXCR_out", "Defined_isotype", "Singlets", "Correct_pairing")), y = Proportion, group = Sample, color = Sample)) +
  geom_point(aes(x = factor(Type, levels = c("MiXCR_out", "Defined_isotype", "Singlets", "Correct_pairing")), y = Proportion, group = Sample, color = Sample)) +
  geom_text(aes(x = factor(Type, levels = c("MiXCR_out", "Defined_isotype", "Singlets", "Correct_pairing")), y = Proportion, group = Sample, color = Sample, label = round(Proportion,2)), hjust = 1.2, check_overlap = TRUE) +
  scale_color_brewer(type = "qual", palette = "Paired") +
  theme_half_open() +
  theme(legend.position = "right", axis.text.x = element_text(angle = 0)) +
  labs(title = "Contigs proportion remaining after processing", x = "Processing step")

```
```{r fig.height=6.5, fig.width=5}
p_count_processing <- ggplot(data = count_processing %>% filter(Type == "Correct_pairing"), aes(x = Sample, y = Proportion)) +
  geom_col(fill = "#E07A6C") +
  geom_text(aes(label = round(Proportion,2)), vjust = -0.2) +
  scale_x_discrete(labels = str_sub(names(sample_all), end = -9L)) +
  scale_y_continuous(breaks = seq(0,1, by = 0.1), limits = c(0,1)) +
  theme_half_open() +
  theme(axis.text.x = element_text(angle = 90)) +
  labs(x = "Sample", y = "Proportion of contigs remaining")
  
p_count_processing

#ggsave("./figures/contig_count_d3.pdf", p_count_processing, height = 6.5, width = 5, dpi = 600)
```

# Collapse the contigs into clonotypes
```{r eval=FALSE, include=FALSE}
# Split the samples into heavy and light chain
# Check how many cells contigs a clonotype has (by cloneID)
lapply(sample_all_paired_2, function(x) table(table(x$cloneId)))

# Collapse contigs into clonotypes by: group the rows by cloneID, paste the cell barcodes together with ;, sum up the readCount and unique UMI count columns, create a new cellCount column by counting the number of rows in each group, the remaining columns SHOULD be identical, and just take the unique value of that.

sample_all_clonotypes <- map(sample_all_paired_2, function (x)
  x %>% group_by(cloneId) %>% 
  summarise(tagValueCELL = paste(tagValueCELL, collapse = ";"),
            readCount = sum(readCount), uniqueUMICount = sum(uniqueUMICount),
            cellCount = length(cloneId), targetSequences = unique(targetSequences),
            targetQualities = unique(targetQualities),
            allVHitsWithScore = unique(allVHitsWithScore),
            allDHitsWithScore = unique(allDHitsWithScore),
            allJHitsWithScore = unique(allJHitsWithScore),
            allCHitsWithScore = unique(allCHitsWithScore),
            allVAlignments = unique(allVAlignments),
            allDAlignments = unique(allDAlignments),
            allJAlignments = unique(allJAlignments),
            allCAlignments = unique(allCAlignments),
            nSeqFR1 = unique(nSeqFR1), minQualFR1 = unique(minQualFR1),
            nSeqCDR1 = unique(nSeqCDR1), minQualCDR1 = unique(minQualCDR1),
            nSeqFR2 = unique(nSeqFR2), minQualFR2 = unique(minQualFR2),
            nSeqCDR2 = unique(nSeqCDR2), minQualCDR2 = unique(minQualCDR2),
            nSeqFR3 = unique(nSeqFR3), minQualFR3 = unique(minQualFR3),
            nSeqCDR3 = unique(nSeqCDR3), minQualCDR3 = unique(minQualCDR3),
            nSeqFR4 = unique(nSeqFR4), minQualFR4 = unique(minQualFR4),
            aaSeqFR1 = unique(aaSeqFR1), aaSeqCDR1 = unique(aaSeqCDR1),
            aaSeqFR2 = unique(aaSeqFR2), aaSeqCDR2 = unique(aaSeqCDR2),
            aaSeqFR3 = unique(aaSeqFR3), aaSeqCDR3 = unique(aaSeqCDR3),
            aaSeqFR4 = unique(aaSeqFR4), refPoints = unique(refPoints),
            bestVGene = unique(bestVGene), bestDGene = unique(bestDGene),
            bestJGene = unique(bestJGene), bestCGene = unique(bestCGene),
            chains = unique(chains), nSeqVDJRegion = unique(nSeqVDJRegion),
            aaSeqVDJRegion = unique(aaSeqVDJRegion),
            aaSeqImputedVDJRegion = unique(aaSeqImputedVDJRegion)))
names(sample_all_clonotypes) <- str_replace_all(names(sample_all_clonotypes), "contigs", "clonotypes")
# Save the full clonotypes
#map2(.x = sample_all_clonotypes, .y = names(sample_all_clonotypes), function (x,y) write_tsv(x, file = paste0("./3_Ig_all_filtered_clonotypes/",y,".txt")))

# Split each sample into heavy and light chain
sample_all_clonotypes_heavy <- map(sample_all_clonotypes, function(x)
  x %>% filter(chains == "IGH"))
sample_all_clonotypes_light <- map(sample_all_clonotypes, function(x)
  x %>% filter(chains != "IGH"))

# Save the split clonotypes
#map2(.x = sample_all_clonotypes_heavy, .y = names(sample_all_clonotypes_heavy), function (x,y) write_tsv(x, file = paste0("./4_Ig_all_clonotypes_split/",y,"_heavy.txt")))
#map2(.x = sample_all_clonotypes_light, .y = names(sample_all_clonotypes_light), function (x,y) write_tsv(x, file = paste0("./unmatched_with_bulkseq/",y,"_light.txt")))
```

# Load with immunarch, Set clone count to cell count, re-recalculate clonal proportion, sort by descending clone size
```{r eval=FALSE, include=FALSE}
data_d3_scseq <- repLoad("./4_Ig_all_clonotypes_split/")

# Load the mixcr files using read_tsv
files_mixcr <- list.files("./4_Ig_all_clonotypes_split/", full.names = T)
Ig_all_mixcr <- map(.x = files_mixcr, function (x) read_tsv(x))

# Replace the clonal count in immunarch with the cell barcode count in mixcr, gene name without score, and imputed VDJ seq (for future cloning)
for (i in 1:length(data_d3_scseq$data)){
  data_d3_scseq$data[[i]]$Clones <- Ig_all_mixcr[[i]]$cellCount
  data_d3_scseq$data[[i]]$V.name <- Ig_all_mixcr[[i]]$bestVGene
  data_d3_scseq$data[[i]]$D.name <- Ig_all_mixcr[[i]]$bestDGene
  data_d3_scseq$data[[i]]$J.name <- Ig_all_mixcr[[i]]$bestJGene
  data_d3_scseq$data[[i]]$C.name <- Ig_all_mixcr[[i]]$bestCGene
  data_d3_scseq$data[[i]]$Sequence.aa <- Ig_all_mixcr[[i]]$aaSeqVDJRegion
  data_d3_scseq$data[[i]]$Sequence.aa.imputed <- Ig_all_mixcr[[i]]$aaSeqImputedVDJRegion
  data_d3_scseq$data[[i]]$UMI.count <- Ig_all_mixcr[[i]]$uniqueUMICount
  data_d3_scseq$data[[i]]$Barcode <- Ig_all_mixcr[[i]]$tagValueCELL
}

# Sort the df by descending number of clones
data_d3_scseq$data <- map(data_d3_scseq$data, function(x) arrange(x, desc(Clones)))

# Rename the isotype to be more concise, recalculate the clonal proportion, get the full-length aa sequence
for (i in 1:length(data_d3_scseq$data)){
  data_d3_scseq$data[[i]]$Proportion <- data_d3_scseq$data[[i]]$Clones/sum(data_d3_scseq$data[[i]]$Clones)}

# Save the processed immunarch object as RDS
saveRDS(data_d3_scseq, "./data_d3_scseq.RDS")
```
